### PR TITLE
net-misc/kea: various improvements

### PR DESCRIPTION
**Pull request by expeditioneer**
fixed-gtest detection
added Systemd services

Closes: https://bugs.gentoo.org/693332
Closes: https://bugs.gentoo.org/626280
Bug: https://bugs.gentoo.org/751883

Signed-off-by: Dennis Lamm <expeditioneer@gentoo.org>
Package-Manager: Portage-3.0.20, Repoman-3.0.3

### CODE REVIEW

File: net-misc/kea/files/kea-1.8.2-gtest.patch [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
++                        GTEST_INCLUDES="-I$dir/include"
++                        GTEST_LDFLAGS="-L$dir/lib64"
++                        GTEST_LDADD="-lgtest"
++                        GTEST_FOUND="true"
++                        break
++                    else
++		        AC_MSG_ERROR([Cannot find gtest in: $GTEST_PATHS])
++		   fi
++	        fi
++	    done
```
[the remark this answers is not in the document]
Comment (expeditioneer) [reply net-misc/kea/files/kea-1.8.2-gtest.patch]: Done

File: net-misc/kea/kea-1.8.2-r1.ebuild [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	econf "${myeconfargs[@]}"
+}
+
+src_install() {
+	default
+	newconfd "${FILESDIR}"/${PN}-confd-r1 ${PN}
+	newinitd "${FILESDIR}"/${PN}-initd-r1 ${PN}
+
+	if use samples; then
+	    diropts -m 0750 -o root -g dhcp
```
[the remark this answers is not in the document]
Comment (expeditioneer) [reply net-misc/kea/kea-1.8.2-r1.ebuild]: Fixed

File: net-misc/kea/kea-1.8.2-r1.ebuild [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+		cp "${FILESDIR}"/kea-dhcp4.conf "${ED}"/etc/kea/kea-dhcp4.conf || die "Could not create kea kea-dhcp4.conf"
+		cp "${FILESDIR}"/kea-dhcp6.conf "${ED}"/etc/kea/kea-dhcp6.conf || die "Could not create kea-dhcp6.conf"
+	fi
+
+	systemd_dounit "${FILESDIR}"/kea-ctrl-agent.service
+	systemd_dounit "${FILESDIR}"/kea-ddns-server.service
+	systemd_dounit "${FILESDIR}"/kea-dhcp4-server.service
+	systemd_dounit "${FILESDIR}"/kea-dhcp6-server.service
+
+	newtmpfiles "${FILESDIR}"/${PN}.tmpfiles.conf ${PN}.conf
```
[the remark this answers is not in the document]
Comment (expeditioneer) [reply net-misc/kea/kea-1.8.2-r1.ebuild:100]: Added missing pkg_postinst

File: net-misc/kea/kea-1.9.10-r1.ebuild [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+	sed -i \
+		-e '/mkdir -p $(DESTDIR)${runstatedir}\/${PACKAGE_NAME}/d' \
+		Makefile.am || die "Fixing Makefile.am failed"
+
+	sed -i \
+		-e 's#test -f "$dir/lib/libgtest.a"#test -f "$dir/lib64/libgtest.a"#g' \
+		-e 's#test -f "$dir/lib/libgtest.so"#test -f "$dir/lib64/libgtest.so"#g' \
+		-e 's GTEST_LDFLAGS="-L$dir/lib GTEST_LDFLAGS="-L$dir/lib64 g' \
+		m4macros/ax_gtest.m4 || die "fixing gtest detection macro failed"
```
[the remark this answers is not in the document]
Comment (expeditioneer) [reply net-misc/kea/kea-1.9.10-r1.ebuild]: Done

**Comment by expeditioneer** [comment]
@Polynomial-C: I've added the gtest detection via pkg-config and squashed the commits

File: net-misc/kea/kea-1.8.2-r1.ebuild [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+	newconfd "${FILESDIR}"/${PN}-confd-r1 ${PN}
+	newinitd "${FILESDIR}"/${PN}-initd-r1 ${PN}
+
+	if use samples; then
+		diropts -m 0750 -o root -g dhcp
+		dodir /etc/kea
+		cp "${FILESDIR}"/kea-ctrl-agent.conf "${ED}"/etc/kea/kea-ctrl-agent.conf || die "Could not create kea-ctrl-agent.conf"
+		cp "${FILESDIR}"/kea-ddns-server.conf "${ED}"/etc/kea/kea-ddns-server.conf || die "Could not create kea-ddns-server.conf"
+		cp "${FILESDIR}"/kea-dhcp4.conf "${ED}"/etc/kea/kea-dhcp4.conf || die "Could not create kea kea-dhcp4.conf"
+		cp "${FILESDIR}"/kea-dhcp6.conf "${ED}"/etc/kea/kea-dhcp6.conf || die "Could not create kea-dhcp6.conf"
```
Comment (Polynomial-C) [inline net-misc/kea/kea-1.8.2-r1.ebuild:92]: Please replace these with `insinto` and `doins` calls. That way you may even remove the `dodir` line.

**Review by Polynomial-C — CHANGES_REQUESTED** [review]

**Comment by Polynomial-C** [comment]
Oh and please fix these issues as well:
```
Traceback (most recent call last):
  File "/usr/lib/portage/python3.9/doins.py", line 594, in <module>
    sys.exit(main(sys.argv[1:]))
  File "/usr/lib/portage/python3.9/doins.py", line 582, in main
    if _doins(
  File "/usr/lib/portage/python3.9/doins.py", line 434, in _doins
    return install_runner.install_file(source, os.path.dirname(dest))
  File "/usr/lib/portage/python3.9/doins.py", line 370, in install_file
    return self._ins_runner.run(source, dest_dir)
  File "/usr/lib/portage/python3.9/doins.py", line 179, in run
    sstat = os.stat(source)
FileNotFoundError: [Errno 2] No such file or directory: b'/var/tmp/portage/net-misc/kea-1.9.10-r1/files/kea-dhcp-ddns-server.service'
 * ERROR: net-misc/kea-1.9.10-r1::gentoo failed (install phase):
 *   doins failed
 *
 * If you need support, post the output of `emerge --info '=net-misc/kea-1.9.10-r1::gentoo'`,
 * the complete build log and the output of `emerge -pqv '=net-misc/kea-1.9.10-r1::gentoo'`.
 * The complete build log is located at '/var/tmp/portage/net-misc/kea-1.9.10-r1/temp/build.log'.
 * The ebuild environment file is located at '/var/tmp/portage/net-misc/kea-1.9.10-r1/temp/environment'.
 * Working directory: '/var/tmp/portage/net-misc/kea-1.9.10-r1/work/kea-1.9.10'
 * S: '/var/tmp/portage/net-misc/kea-1.9.10-r1/work/kea-1.9.10'
 * QA Notice: Unrecognized configure options:
 *
 *      --with-gtest
 *      --with-gtest
```

**Comment by expeditioneer** [comment]
@Polynomial-C this should be fixed now

**Comment by expeditioneer** [comment]
Added bugs to commit message

File: net-misc/kea/kea-1.8.2-r1.ebuild [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+	newconfd "${FILESDIR}"/${PN}-confd-r1 ${PN}
+	newinitd "${FILESDIR}"/${PN}-initd-r1 ${PN}
+
+	if use samples; then
+		diropts -m 0750 -o root -g dhcp
+		insinto /etc/kea
+		doins "${FILESDIR}"/${PN}-ctrl-agent.conf || die "Could not create kea-ctrl-agent.conf"
+		doins "${FILESDIR}"/${PN}-ddns-server.conf || die "Could not create kea-ddns-server.conf"
+		doins "${FILESDIR}"/${PN}-dhcp4.conf || die "Could not create kea kea-dhcp4.conf"
+		doins "${FILESDIR}"/${PN}-dhcp6.conf || die "Could not create kea-dhcp6.conf"
```
Comment (Polynomial-C) [inline net-misc/kea/kea-1.8.2-r1.ebuild:93]: `doins` does not require `|| die` calls. See https://devmanual.gentoo.org/ebuild-writing/error-handling/index.html

File: net-misc/kea/kea-1.8.2-r1.ebuild [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+	econf "${myeconfargs[@]}"
+}
+
+src_install() {
+	default
+	newconfd "${FILESDIR}"/${PN}-confd-r1 ${PN}
+	newinitd "${FILESDIR}"/${PN}-initd-r1 ${PN}
+
+	if use samples; then
+		diropts -m 0750 -o root -g dhcp
```
Comment (Polynomial-C) [inline net-misc/kea/kea-1.8.2-r1.ebuild:88]: Alright, removing the `dodir` call also made this line useless. But I suppose you want the `etc/kea` dir to have these permissions so I suggest to re-introduce the `dodir /etc/kea` call.

**Review by Polynomial-C — CHANGES_REQUESTED** [review]

File: net-misc/kea/kea-1.8.2-r1.ebuild (hunk 6 above)
Comment (expeditioneer) [reply net-misc/kea/kea-1.8.2-r1.ebuild]: Done

File: net-misc/kea/kea-1.8.2-r1.ebuild (hunk 7 above)
Comment (expeditioneer) [reply net-misc/kea/kea-1.8.2-r1.ebuild:88]: I've fixed it

**Comment by expeditioneer** [comment]
@Polynomial-C could we merge this?

File: net-misc/kea/files/kea-1.8.2-gtest.patch [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+-        if test "$GTEST_VERSION" == "googletest"; then
+-            GTEST_VERSION=${GTEST_SOURCE%"/googletest"}
+-            GTEST_VERSION=`basename $GTEST_VERSION`
+-        fi
+-        GTEST_VERSION="${GTEST_VERSION#googletest-release-}"
+-        GTEST_VERSION="${GTEST_VERSION#gtest-}"
+-    fi
++    DISTCHECK_GTEST_CONFIGURE_FLAG="--with-gtest"
++    PKG_CHECK_MODULES([GTEST], [gtest], [], [AC_MSG_ERROR([gtest requested but not found])])
++    GTEST_INCLUDES=`pkg-config --keep-system-cflags --cflags-only-I gtest`
```
Comment (thesamesam) [inline net-misc/kea/files/kea-1.8.2-gtest.patch:82]: You shouldn't need to then call pkg-config directly (PKG_CHECK_MODULES should populate everything you need), but if you absolutely must, respect ${PKG_CONFIG} instead (not calling just pkg-config bare).

Comment (expeditioneer) [reply net-misc/kea/files/kea-1.8.2-gtest.patch]: @thesamesam thank you for the hint with ${PKG_CONFIG} i replaced  the plain `pkg-config`calls. Sadly PKG_CHECK_MODULES does not provide everything which is required by kea. I also wanted to keep the changes minimal on their codebase.

**Comment by expeditioneer** [comment]
I've found an issue with the conditional activation of gtest. Please wait with further reviews until i have fixed that.

**Comment by expeditioneer** [comment]
The conditional activation of gtest should now work as expected and intended.

File: net-misc/kea/kea-1.9.10-r1.ebuild [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+src_install() {
+	default
+	newconfd "${FILESDIR}"/${PN}-confd-r1 ${PN}
+	newinitd "${FILESDIR}"/${PN}-initd-r1 ${PN}
+
+	if use samples; then
+		diropts -m 0750 -o root -g dhcp
+		dodir /etc/kea
+        insopts -m 0640 -o root -g dhcp
```
Comment (Polynomial-C) [inline net-misc/kea/kea-1.9.10-r1.ebuild:89]: Please fix the indentation

File: net-misc/kea/kea-9999.ebuild [hunk 10]
```diff
@@ -65,7 +74,29 @@ src_install() {
 	default
 	newconfd "${FILESDIR}"/${PN}-confd-r1 ${PN}
 	newinitd "${FILESDIR}"/${PN}-initd-r1 ${PN}
-	keepdir /var/lib/${PN} /var/log
-	rm -rf "${ED}"/var/run || die
+
+	if use samples; then
+		diropts -m 0750 -o root -g dhcp
+		dodir /etc/kea
+        insopts -m 0640 -o root -g dhcp
```
Comment (Polynomial-C) [inline net-misc/kea/kea-9999.ebuild:81]: Likewise

**Review by Polynomial-C — CHANGES_REQUESTED** [review]
Please fix the two remaining minor issues.
**Optionally** you may also bump kea to version `1.9.11` (but in a separate commit please).

Once you've done this, you can commit your PR. Thanks for your patience.

File: net-misc/kea/kea-9999.ebuild (hunk 10 above)
Comment (expeditioneer) [reply net-misc/kea/kea-9999.ebuild]: done

File: net-misc/kea/kea-1.9.10-r1.ebuild (hunk 9 above)
Comment (expeditioneer) [reply net-misc/kea/kea-1.9.10-r1.ebuild]: done

File: net-misc/kea/files/kea-1.8.2-gtest.patch (hunk 8 above)
Comment (thesamesam) [reply net-misc/kea/files/kea-1.8.2-gtest.patch]: np, I understand -- thanks! I defer to @Polynomial-C now.

**Pull request closed without merging**